### ISSUE TITLE
Add secret description to metadata

### ISSUE THREAD

**Issue by sgutwein** [hashicorp/vault#10714]
**Is your feature request related to a problem? Please describe.**
Assume you have a secret vault with a ton of secrets. The secrets are only declared by the folder and the secret name. At one point it is very difficult to have an overview, where the secrets are used, where are they came from and for which reason.

**Describe the solution you'd like**
It would be very nice have some kind of "description field" in the secret metadata. If a user has access to the metadata he should be able to see the descriptions/notes.

**Describe alternatives you've considered**
Today (with Vault 1.6.1) I create a key named `notes` in the secret and add further information in the assigned value. 
This is not a good solution as you need access rights to read the data / values of the secret, to see the notes.

**Comment by aphorise** [comment]
@sgutwein can you not already accomplish this [using KV](https://www.vaultproject.io/docs/secrets/kv/kv-v2) for example:

```
vault secrets enable -version=2 kv ;
vault kv put kv/a value='my secret value here' metadata1="was created at: $(date)" metadata2="by user: $(whoami)" ;
vault kv get -field=metadata1 kv/a ;
vault kv get -field=metadata2 kv/a ;
vault kv get -format=json kv/a ;
```

**Comment by sgutwein** [comment]
@aphorise Is it possible to access and edit those Values by UI?

**Comment by aphorise** [comment]
Sure - see for example:
<img width="875" alt="Vault_Dev_Mode" src="https://user-images.githubusercontent.com/974854/104906443-fde28200-5983-11eb-9dec-aac47d2fc76c.png">

**Comment by sgutwein** [comment]
Aha, I think this is what I described in "alternatives you've considered". 

You can't access those values without **data reading rights** (as the values of metadata1/metadata2 are handled as secrets by Vault). This not a good solution, I want a description which can be accessed with **metadata reading rights**. Or did I missed something? If that's the case - sry! :)

I was thinking of something like this: https://ibb.co/Wvx6RqX

**Comment by aphorise** [comment]
@sgutwein - maybe open a feature-request for [supporting `allowed_parameters` and or `required_parameters`](https://www.vaultproject.io/docs/concepts/policies#parameter-constraints) - as part of reads which might suit what you're after; that way you could use the same policy as exemplified in the docs (which works for writes atm) and be able to perform a `vault kv get...` that's scoped to a `read` only policy along the same lines.

**Comment by sgutwein** [comment]
@aphorise I don't think a description as a `parameter / value` pair is a good solution, as parameters can only accessed by `data` read rights (I'm referring to kv2). In my opinion the description of a secret should be in the [metadata](https://www.vaultproject.io/docs/commands/kv/metadata) of the secret. Please let me explain:
 
This is an example of the current [metadata](https://www.vaultproject.io/docs/commands/kv/metadata):
```
$ vault kv metadata get secret/creds
========== Metadata ==========
Key                     Value
---                     -----
cas_required            false
created_time            2019-06-28T15:53:30.395814Z
current_version         5
delete_version_after    0s
max_versions            0
oldest_version          0
updated_time            2019-06-28T16:01:47.40064Z

====== Version 1 ======
Key              Value
---              -----
created_time     2019-06-28T15:53:30.395814Z
deletion_time    n/a
destroyed        false

====== Version 2 ======
Key              Value
---              -----
created_time     2019-06-28T16:01:36.676912Z
deletion_time    n/a
destroyed        false
```

I was thinking of adding a field  called `description` to the metadata, which is as default (for backwards compatibility) empty:
```
$ vault kv metadata get secret/creds
========== Metadata ==========
Key                     Value
---                     -----
cas_required            false
created_time            2019-06-28T15:53:30.395814Z
current_version         5
delete_version_after    0s
max_versions            0
oldest_version          0
updated_time            2019-06-28T16:01:47.40064Z

====== Version 1 ======
Key              Value
---              -----
**description ""**
created_time     2019-05-28T15:53:30.395814Z
deletion_time    n/a
destroyed        false

====== Version 2 ======
Key              Value
---              -----
**description "This secret is used by xyz to gain access to abc"**
created_time     2021-06-28T15:53:30.395814Z
deletion_time    n/a
destroyed        false

====== Version 3 ======
Key              Value
---              -----
**description "This secret is used by xyz to gain access to abc1234new"**
created_time     2021-06-28T16:01:36.676912Z
deletion_time    n/a
destroyed        false
```

As user with these rights should be able to access and edit them (also on the UI):
```
path "xyz/metadata/*" {
    capabilities =  ["read", "list", "update"]
}

path "xyz/data/*" {
    capabilities =  ["create", "update"]
}
```

I hope it is now clear what I'm up to! :)

**Comment by ccapurso** [comment]
Thank you for your contribution, @sgutwein! The 1.9 release will include an addition of a `custom_metadata` field to the key metadata for a secret in version 2 of the KV secret engine. This field allows for arbitrary string to string key-value pairs that can be used to provide descriptive metadata to your secret entries. One thing to note is that a secret's key metadata is version-agnostic.

The Vault team has recently released the first release candidate of Vault 1.9. Please see this list of the [available binaries](https://releases.hashicorp.com/vault/1.9.0-rc1/).

You can also view the pending doc updates here:
https://github.com/hashicorp/vault/blob/release/1.9.x/website/content/docs/secrets/kv/kv-v2.mdx
https://github.com/hashicorp/vault/blob/release/1.9.x/website/content/api-docs/secret/kv/kv-v2.mdx

Do you think an addition like this provides a solution that fits your needs?
